Centralize the code that figures out which memory ranges to save into core files

clayborg opened the pull request:
Prior to this patch, each core file plugin (ObjectFileMachO.cpp and ObjectFileMinindump.cpp) would calculate the address ranges to save in different ways. This patch adds a new function to Process.h/.cpp:

```
Status Process::CalculateCoreFileSaveRanges(lldb::SaveCoreStyle core_style, CoreFileMemoryRanges &ranges);
```

The patch updates the ObjectFileMachO::SaveCore(...) and ObjectFileMinindump::SaveCore(...) to use same code. This will allow core files to be consistent with the lldb::SaveCoreStyle across different core file creators and will allow us to add new core file saving features that do more complex things in future patches.

clayborg committed:
Centralize the code that figures out which memory ranges to save into core files.

Prior to this patch, each core file plugin (ObjectFileMachO.cpp and ObjectFileMinindump.cpp) would calculate the address ranges to save in different ways. This patch adds a new function to Process.h/.cpp:

```
Status Process::CalculateCoreFileSaveRanges(lldb::SaveCoreStyle core_style, CoreFileMemoryRanges &ranges);
```

The patch updates the ObjectFileMachO::SaveCore(...) and ObjectFileMinindump::SaveCore(...) to use same code. This will allow core files to be consistent with the lldb::SaveCoreStyle across different core file creators and will allow us to add new core file saving features that do more complex things in future patches.

jasonmolenda commented on lldb/include/lldb/Target/Process.h:
```diff
@@ -704,7 +705,37 @@ class Process : public std::enable_shared_from_this<Process>,
+  using CoreFileMemoryRanges = std::vector<CoreFileMemoryRange>;
+
+  /// Helper function for Process::SaveCore(...) that calculates the address
+  /// ranges that should be save. This allows all core file plug-ins to save
```
"should be saved"

jasonmolenda commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+  }
+
+  // Also check with our threads and get the regions for their stack pointers
+  // and add those regions if not already added above
```
This is a clever idea.  Add a period at the end of the comment.

jasonmolenda commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+  // and add those regions if not already added above
+  ThreadList &thread_list = process.GetThreadList();
+  const size_t num_threads = thread_list.GetSize();
+  for (size_t idx=0; idx<num_threads; ++idx) {
```
ThreadList has an interable interface, couldn't you do `for (ThreadSP &thread_sp : thread_list.Threads())`?  I haven't actually tried this.

jasonmolenda reviewed (approved):
A couple of small nits but looks good to me.  I do have a little concern about pulling in `llvm/ADT/AddressRanges.h` in Process.h to get llvm::AddressRange when we have an lldb_private::AddressRange, which is used in a number of places in lldb.  I'm surprised you only needed to add `lldb_private::` namespace qualifications in that TraceDumper after adding the llvm header in Process.h.  I don't have a good suggestion here but it seems like the kind of thing that's going to annoy someone once a year or so when they stumble on the ambiguity. :)

clayborg commented on lldb/include/lldb/Target/Process.h:
```diff
@@ -704,7 +705,37 @@ class Process : public std::enable_shared_from_this<Process>,
+  /// Helper function for Process::SaveCore(...) that calculates the address
+  /// ranges that should be save. This allows all core file plug-ins to save
+  /// consistent memory ranges given a \a core_style.
+  Status CalculateCoreFileSaveRanges(lldb::SaveCoreStyle core_style,
+                                     CoreFileMemoryRanges &ranges);
+
```
I did that initially but then it was always just converted to Status in the end so it seemed like using Expected when we didn't need to. I am happy to change to Expected<> if you really think it should be that way

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+  return {range, region.GetLLDBPermissions()};
+}
+
+// Add dirty pages to the core file ranges and return true if dirty pages
+// were added. Return false if the dirty page information is not valid or in
+// the region.
+static bool AddDirtyPages(const MemoryRegionInfo &region,
```
We aren't going to propagate the error, we just need to know if it succeeded or not, so I didn't have it return an error. Otherwise every single range from a linux lldb-server will return an error "dirty pages not supported". For Darwin we added special extra features to track dirty pages, but no one else has these. Again, I can add an error here, but I will just end up consuming the error because lack of dirty page information isn't going to stop a core file from being created. Also many regions on mac will not have any dirty pages if they don't have write permissions, so most regions would end up returning an error, which again, we won't do anything with, we will just ignore.

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+// Create a CoreFileMemoryRange from a MemoryRegionInfo
+static Process::CoreFileMemoryRange
+CreateCoreFileMemoryRange(const MemoryRegionInfo &region) {
+  const addr_t addr = region.GetRange().GetRangeBase();
+  llvm::AddressRange range(addr, addr + region.GetRange().GetByteSize());
```
This is an llvm::AddressRange here. We also have lldb_private::AddressRange which contains a lldb_private::Address (section offset address) + byte size. The llvm::AddressRange just has a start address and end address. I could add accessors to MemoryRegionInfo for these, but as Jason already eluded to we have llvm::AddressRange and lldb_private::AddressRange and if we include both header files in MemoryRegionInfo.h, we might end up with more qualifications needed on bare `AddressRange` types in source files.

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+// given region. If the region has dirty page information, only dirty pages
+// will be added to \a ranges, else the entire range will be added to \a
+// ranges.
+static void AddRegion(const MemoryRegionInfo &region, bool try_dirty_pages,
```
`try_dirty_pages` is saying "if you can add only the dirty pages from the current region because the process plugin supports it, then please add those, if not please add the full region. If you note in the code below we only try to add dirty pages if we are asked to when `try_dirty_pages == true`

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+// given region. If the region has dirty page information, only dirty pages
+// will be added to \a ranges, else the entire range will be added to \a
+// ranges.
+static void AddRegion(const MemoryRegionInfo &region, bool try_dirty_pages,
+                      Process::CoreFileMemoryRanges &ranges) {
```
If we have empty ranges or a memory region without permissions, we don't need to add it to the core file, and this function will determine if it needs to add the region to the list of regions to save to the core file. The user won't care or need to know about any of the error conditions.

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+  const bool try_dirty_pages = false;
+  for (const auto &region: regions)
+    AddRegion(region, try_dirty_pages, ranges);
+  return Status();
```
Each of the `GetCoreFileSaveRanges*` funtions all have the same return value to keep it consistent. When we ask explicitly for dirty pages only with the `eSaveCoreDirtyOnly` core style, that function returns an error if dirty pages are not supported by any regions. Not much can go wrong in this function. I can remove the Status return value if needed.

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -6252,3 +6243,188 @@ Status Process::WriteMemoryTags(lldb::addr_t addr, size_t len,
+// Save all thread stacks to the core file. Some OS versions support reporting
+// when a memory region is stack related. We check on this information, but we
+// also use the stack pointers of each thread and add those in case the OS
+// doesn't support reporting stack memory. This function does unique the stack
+// regions and won't add the same range twice. This function also attempts to
```
sounds good

clayborg commented:
I addressed most of the feedback. Alex let me know if you still really want llvm::Error and llvm::Expected to be used as I can add that if you think it is required. I also ran clang format.

clayborg commented:
@jasonmolenda I was wondering if we should modify `GetCoreFileSaveRangesDirtyOnly(...)` to try and add all dirty pages and see if any regions have the dirty page info, but if no memory region infos have the dirty pages information, then fall back to adding all memory regions with `write` access. What do you think? This would allow systems to not support dirty pages, but get a more minimal core file saved out. The other option is to add a new `lldb::SaveCoreStyle` enum like `lldb::SaveCoreStyle::eSaveCoreWriteOnly` to save only sections that have write permissions.

Right now we have `eSaveCoreFull` which saves all memory regions that have any valid permissions, `eSaveCoreDirtyOnly` which will currently only save dirty pages if it is supported by the process plugin (ProcessGDBRemove for most people, where `debugserver` supports the page stuff, but `lldb-server` doesn't), and `eSaveCoreStackOnly` which only emits the stacks (where it will use the dirty page info if available and fall back to saving the entire stack range if dirty page support is not available).

clayborg commented:
@bulbazord let me know if you require any changes after reading my inline comments.

bulbazord commented:
> I addressed most of the feedback. Alex let me know if you still really want llvm::Error and llvm::Expected to be used as I can add that if you think it is required. I also ran clang format.

I think your answers make sense to me. I don't think you need to add them here since it all eventually turns into Status objects anyway... I generally prefer it over Status though. Thanks for checking. 👍

jasonmolenda commented:
> @jasonmolenda I was wondering if we should modify `GetCoreFileSaveRangesDirtyOnly(...)` to try and add all dirty pages and see if any regions have the dirty page info, but if no memory region infos have the dirty pages information, then fall back to adding all memory regions with `write` access. What do you think? 

Another good idea.  Do we get page permissions like that on Linux?  I'm curious what system it would work on.

I tried a full memory coredump on darwin and there ARE some regions that are marked `r` or `rx` and yet have a dirty page.  Some of them are tagged as malloc-metadata so maybe the page protections were flipped to avoid buggy overwriting by the process or something, I didn't look too closely into them, it was only a dozen pages of memory total.  So at least on darwin, if debugserver didn't provide the dirty-pages list for each memory region, it'd probably be indistinguishable by developers who are mostly wanting to get stack+heap with these things.

clayborg committed:
Make changes according to feedback

- If we are saving core style with eSaveCoreDirtyOnly and process doesn't support dirty ranges, save all regions with write permssions as a backup.
- Modify the TestProcessSaveCoreMinidump.py test case to handle minidump now able to handle eSaveCoreDirtyOnly and eSaveCoreFull

clayborg committed:
Make eSaveCoreDirtyOnly save all regions with write permissions if no dirty page support is detected and do full testing on minidumps.

antmox commented:
Hello @clayborg ! Looks like this patch broke lldb-aarch64-windows bot:  https://lab.llvm.org/buildbot/#/builders/219/builds/6868
Could you please look at this ?

clayborg commented:
> Hello @clayborg ! Looks like this patch broke lldb-aarch64-windows bot: https://lab.llvm.org/buildbot/#/builders/219/builds/6868 Could you please look at this ?

@antmox 

I looked at this and I didn't touch anything related to native windows core file saving. If this is a native windows build on arm64 for lldb, and we are debugging with COFF files, then this code will get run:
```
bool ObjectFilePECOFF::SaveCore(const lldb::ProcessSP &process_sp,
                                const lldb_private::FileSpec &outfile,
                                lldb::SaveCoreStyle &core_style,
                                lldb_private::Status &error) {
  core_style = eSaveCoreFull;
  return SaveMiniDump(process_sp, outfile, error);
}
```
The `SaveMinidump` calls native windows APIs to create the minidump file in `lldb/source/Plugins/ObjectFile/PECOFF/WindowsMiniDump.cpp`. Then this file gets loaded by the ProcessMinidump.cpp and I didn't touch anything inside of the process plug-in that loads the minidumps. So I am not sure how to proceed other than disabling this test for arm64 windows? Does anyone have access to an arm64 windows machine to see what is going on? In theory none of the code I changed should touch this, and I didn't modify `lldb/test/API/functionalities/postmortem/minidump/TestMiniDump.py` at all with this patch.

clayborg commented:
Does anyone have access to a native windows arm64 build of LLDB that could help look at this?

antmox commented:
@clayborg 
I will try to reproduce on the buildbot windows machine. build ongoing

clayborg commented:
> @clayborg I will try to reproduce on the buildbot windows machine. build ongoing

If you end up being able to reproduce, if you can send the minidump file that was produced I should be able to look at it to make sure it was not produced by the core file saver. But if this is debugging a COFF file, it should only be using the native windows API to produce a core file and then load it. And I don't see any changes in this PR that should affect this, but you never know!

antmox commented:
[core_dmps.tar.gz](https://github.com/llvm/llvm-project/files/13355872/core_dmps.tar.gz)
here is an archive with the 2 core.dmp files

clayborg commented:
> [core_dmps.tar.gz](https://github.com/llvm/llvm-project/files/13355872/core_dmps.tar.gz) here is an archive with the 2 core.dmp files

It does indeed seem like my new core file creation code is being triggered on windows. I am working on a fix that adds ARM64 support to the built in minidump creator.

clayborg commented:
> [antmox](/antmox)

https://github.com/llvm/llvm-project/pull/72315

A new PR to fix this issue. Is there any way you can test this to see if this works?

antmox commented:
@clayborg yes sure I will test this on the same machine
